cpp/src/io/orc/writer_impl.cu

```diff
@@ -1186,20 +1196,93 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
+    cudaMemcpyAsync(stat_chunks.data() + chunk_offset,
+                    incoming_stats.stripe_stat_chunks[i].data(),
+                    chunk_bytes,
+                    cudaMemcpyDeviceToDevice,
+                    stream);
+    cudaMemcpyAsync(stats_merge.device_ptr() + merge_offset,
+                    incoming_stats.stripe_stat_merge[i].device_ptr(),
+                    merge_bytes,
+                    cudaMemcpyDeviceToDevice,
+                    stream);
+    chunk_offset += incoming_stats.stripe_stat_chunks[i].size();
+    merge_offset += incoming_stats.stripe_stat_merge[i].size();
+  }
+
+  if (single_write_mode) {
+    std::vector<statistics_merge_group> file_stats_merge(num_file_blobs);
+    for (auto i = 0; i < num_columns; ++i) {
+      auto col_stats         = &file_stats_merge[i];
+      col_stats->col         = incoming_stats.stats_desc.device_ptr(i);
+      col_stats->start_chunk = static_cast<uint32_t>(i * num_stripes);
+      col_stats->num_chunks  = static_cast<uint32_t>(num_stripes);
+    }
+
+    auto d_file_stats_merge = stats_merge.device_ptr(num_stripe_blobs);
+    cudaMemcpyAsync(d_file_stats_merge,
+                    file_stats_merge.data(),
+                    num_file_blobs * sizeof(statistics_merge_group),
+                    cudaMemcpyHostToDevice,
+                    stream);
+
+    auto file_stat_chunks = stat_chunks.data() + num_stripe_blobs;
+    detail::merge_group_statistics<detail::io_file_format::ORC>(
+      file_stat_chunks, stat_chunks.data(), d_file_stats_merge, num_file_blobs, stream);
+  }
+
+  hostdevice_vector<uint8_t> blobs =
+    allocate_and_encode_blobs(stats_merge, stat_chunks, num_blobs, stream);
+
+  auto stripe_stat_merge = stats_merge.host_ptr();
+
   std::vector<ColStatsBlob> stripe_blobs(num_stripe_blobs);
   for (size_t i = 0; i < num_stripe_blobs; i++) {
     auto const stat_begin = blobs.host_ptr(stripe_stat_merge[i].start_chunk);
     auto const stat_end   = stat_begin + stripe_stat_merge[i].num_chunks;
     stripe_blobs[i].assign(stat_begin, stat_end);
   }
 
-  std::vector<ColStatsBlob> file_blobs(num_file_blobs);
-  for (size_t i = 0; i < num_file_blobs; i++) {
-    auto const stat_begin = blobs.host_ptr(file_stat_merge[i].start_chunk);
-    auto const stat_end   = stat_begin + file_stat_merge[i].num_chunks;
-    file_blobs[i].assign(stat_begin, stat_end);
+  std::vector<ColStatsBlob> file_blobs(single_write_mode ? num_file_blobs : 0);
+  if (single_write_mode) {
+    auto file_stat_merge = stats_merge.host_ptr(num_stripe_blobs);
+    for (auto i = 0; i < num_file_blobs; i++) {
+      auto const stat_begin = blobs.host_ptr(file_stat_merge[i].start_chunk);
+      auto const stat_end   = stat_begin + file_stat_merge[i].num_chunks;
+      file_blobs[i].assign(stat_begin, stat_end);
+    }
   }
-  return {std::move(rowgroup_blobs), std::move(stripe_blobs), std::move(file_blobs)};
+
+  return {std::move(stripe_blobs), std::move(file_blobs)};
```

Comment:
It was, but since row groups are needed earlier than when we call `finish_statistics_blobs` it isn't in this structure anymore. Row group information is consumed in write still.

Reply:
Missed that part :( Might be good to rename `encoded_statistics`, since it only includes the ones that are written in the file footer.

Reply:
A great idea, I'm thinking `encoded_footer_statistics`